datastructures/src/ddsketch/atomic.rs

```
@@ -0,0 +1,280 @@
+    pub fn count(&self) -> u64 {
+        self.count.load(Ordering::Relaxed)
+    }
+
+    /// The number of samples that were over the limit and are too
+    /// high to store in any given bucket.
+    pub fn too_high(&self) -> u64 {
+        if self.limit == std::u64::MAX {
+            return 0;
+        }
+
+        self.buckets.last().unwrap().get().into()
+    }
+
+    /// Returns the approximate value of the quantile specified from
+    /// 0.0 to 1.0.
+    ///
+    /// Any value returned that is within the range [0, limit] will be
+    /// accurate within a relative error of `alpha` provided that no
+    /// counters within the sketch have saturated.
+    ///
+    /// If those two conditions are not met, then no error bounds are
+    /// given for the returned quantile.
+    pub fn quantile(&self, q: f64) -> u64 {
+        if q.is_nan() {
+            return 0;
+        }
+
+        let min = self.min.load(Ordering::Relaxed);
+        let max = self.max.load(Ordering::Relaxed);
+
+        if q < 0.0 {
+            return min;
+        }
+        if q >= 1.0 {
+            return max;
+        }
+
+        let rank = (q * self.count() as f64) as u64;
+        let index = self
+            .buckets
+            .iter()
+            .scan(0u64, |total, count| {
+                *total += count.load(Ordering::Relaxed).into();
+                Some(*total)
+            })
+            .enumerate()
+            .skip_while(|&(_, count)| count <= rank)
+            .map(|(i, _)| i)
+            .next();
+
+        let index = match index {
+            Some(idx) if idx < self.cutoff => idx,
+            Some(idx) if idx == self.buckets.len() - 1 => return max,
+            Some(idx) => idx - 1,
+            None => return max,
+        };
+
+        ((self.gamma.powi(index as i32) / (0.5 * (self.gamma + 1.0))).round() as u64)
+            .min(max)
+            .max(min)
+    }
+
+    /// Merge two different sketches.
+    ///
+    /// # Panics
```

Panic is probably fine for now, but we could return an error instead?